// ory-editor-plugins-parallax-background/lib/Component/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var ParallaxBackground = /** @class */ (function (_super) {
    __extends(ParallaxBackground, _super);
    function ParallaxBackground(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {};
        _this.handleDarkenPreviewChange = _this.handleDarkenPreviewChange.bind(_this);
        _this.commitDarken = _this.commitDarken.bind(_this);
        _this.handleBackgroundPreviewChange = _this.handleBackgroundPreviewChange.bind(_this);
        _this.commitBackground = _this.commitBackground.bind(_this);
        return _this;
    }
    ParallaxBackground.prototype.render = function () {
        var Controls = this.props.Controls;
        return (React.createElement(Controls, __assign({}, this.props, { state: __assign({}, this.props.state, { darken: this.state.darken
                    ? this.state.darken
                    : this.props.state.darken, background: this.state.background
                    ? this.state.background
                    : this.props.state.background }), handleDarkenPreviewChange: this.handleDarkenPreviewChange, commitDarken: this.commitDarken, handleBackgroundPreviewChange: this.handleBackgroundPreviewChange, commitBackground: this.commitBackground })));
    };
    ParallaxBackground.prototype.handleDarkenPreviewChange = function (darken) {
        this.setState({ darken: darken });
    };
    ParallaxBackground.prototype.commitDarken = function () {
        this.props.onChange({ darken: this.state.darken });
        this.setState({ darken: undefined });
    };
    ParallaxBackground.prototype.handleBackgroundPreviewChange = function (background) {
        this.setState({ background: background });
    };
    ParallaxBackground.prototype.commitBackground = function () {
        this.props.onChange({ background: this.state.background });
        this.setState({ background: undefined });
    };
    return ParallaxBackground;
}(React.Component));
exports.default = ParallaxBackground;
